### PR TITLE
[839] Sync trainee degrees to DTTP

### PR DESCRIPTION
**Pull request by jhenderson**
### Context
https://trello.com/c/BZsypfif/839-l-multiple-degrees-to-dttp

### Changes proposed in this pull request
- Replace `HESA_DEGREE_TYPES` with `Dttp::CodeSets::DegreeTypes::MAPPING`
- New objects: `Dttp::Params::Contact`, `Dttp::Params::PlacementAssignment`, and `Dttp::Params::DegreeQualification`
- Remove `Dttp::TraineePresenter` (it was getting too big) and replace it with objects for each DTTP entity (easier to understand and manage)
- Update `Dttp::RegisterForTrn` to use the new DTTP entity objects and include trainee degree data in batch request
- Update `Dttp::ContactUpdate` to use the new DTTP entity objects
- Minor refactorings

### Guidance to review
Try it out manually:

```ruby
require_relative "config/environment"
require "factory_bot_rails"
require "faker"

FactoryBot.find_definitions rescue nil

trainee = FactoryBot.create(:trainee, :with_programme_details, :diversity_disclosed, :with_placement_assignment)

trainee.degrees << FactoryBot.create(:degree, :uk_degree_with_details)
trainee.degrees << FactoryBot.create(:degree, :non_uk_degree_with_details)

entity_ids = Dttp::RegisterForTrn.call(trainee: trainee, trainee_creator_dttp_id: "98c63edb-b0a6-e811-812c-5065f38bc301")

puts entity_ids
```

### CODE REVIEW

**Comment by jhenderson** [comment]
[[L] Multiple degrees to DTTP](https://trello.com/c/BZsypfif/839-l-multiple-degrees-to-dttp)

File: app/lib/dttp/entity/contact.rb [hunk 1]
```diff
@@ -0,0 +1,85 @@
+# frozen_string_literal: true
+
+module Dttp
+  module Entity
```
Comment (jhenderson) [inline app/lib/dttp/entity/contact.rb:4]: I'm thinking of renaming `Entity` to 'Params`. Let me know if this is a good idea.

File: app/lib/dttp/entity.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
+
+module Dttp
+  module Entity
+    class << self
```
Comment (gpeng) [inline app/lib/dttp/entity.rb:5]: Do these need to be class methods?

File: app/lib/dttp/entity/degree_qualification.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        params.to_json
+      end
+
+    private
+
+      def build_parms
+        {
+          "dfe_ContactId@odata.bind" => "$#{contact_change_set_id}",
+          "dfe_TrainingRecordId@odata.bind" => "$#{placement_assignment_change_set_id}",
+          "dfe_ClassofDegreeId@odata.bind" => "/dfe_classofdegrees(#{Entity.degree_class_id(degree.grade)})",
```
Comment (gpeng) [inline app/lib/dttp/entity/degree_qualification.rb:25]: Harking back to the 'class methods' comment, we could do away with the `Entity` reference here (and all the other places) if they were instance methods. I think these are only called within `Entity`-ses as far as I can see.

**Review by gpeng — COMMENTED** [review]
This looks great. Good work. Splitting the TraineePresenter up makes things much simpler 👍 Couple of little question/comments but on the whole all looking good.

**Review by gpeng — APPROVED** [review]

**Pull request merged**